al_bench/factory.py

```diff
@@ -0,0 +1,138 @@
+# ==========================================================================
+#
+#   Copyright NumFOCUS
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#          https://www.apache.org/licenses/LICENSE-2.0.txt
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ==========================================================================
+
+from __future__ import annotations
+import numpy as np
+import scipy.stats
+from numpy.typing import NDArray
+from typing import Mapping, MutableMapping, Sequence, Any
+
+
+class ComputeCertainty:
+    def __init__(self, certainty_type, percentiles, cutoffs):
+        """
+        certainty_type can be "negative_entropy", "confidence", or "margin", or a list
+        (or tuple) of one or more of these.  A value of None means all certainty types.
+
+        percentiles is a list (or tuple) of percentile values to be computed, e.g., (5,
+        10, 25, 50).  The P percentile score will be that score S such that P percent of
+        all scores are less than (more uncertain than) S.
+
+        cutoffs: if the `certainty_type` is a single type then `cutoffs` can be a list
```

Comment:
Different types have different ranges, which means that the values to pass to cutoffs varies by type.  Should we normalize the range of the computed values in some manner?  If we think the cdf value would be a useful indicator of whether we have trained enough, then this might be desired.

Reply:
Confidence and margin already naturally fall into a `[0,1]` range, but we could do an affine transformation to negative entropy to get the same range.  That merits further discussion.  However, I worry that that would make it seem like, say,  0.5 is about the same regardless of whether it was confidence, margin, or (normalized) negative entropy.  I don't think it is; 0.5 is harder to achieve as a margin than as a confidence, and normalized negative entropy likewise likely indicates a qualitatively different achievement hardness.

Reply:
It really is trying to predict which cdf values (if any) would be of utility to show a user to help them decide when things were labelled to their satisfaction.  I don't know the right answer, but that doesn't mean we should use a misleading answer.

Reply:
I don't know the right answer either.  In the meantime, a user who chooses to use, say, `negative_entropy`, will have to learn how to interpret the cutoff values.